### ISSUE TITLE
Need to change Column (Nullable ...) to NullableColumn ...

### ISSUE THREAD

**Issue by tomjaguarpaw** [tomjaguarpaw/haskell-opaleye#97]
I realised that `Column (Nullable ...)` doesn't make sense.  Not only does it allow the nonsensical type `Column (Nullable (Nullable ...))` it also means that the type `Column a -> Column a -> Column PGBool` is uncorrect.  It should be `Column a -> Column a -> Column (Nullable PGBool)`.  This is too painful to contemplate!

The correct fix is to have two type constructors, `Column` and `NullableColumn`.  This will break a lot of code.  I propose the following
- 0.5: Add `NullableColumn a` as type alias for `Column (Nullable a)`
- 0.6: Deprecate `Nullable`
- 0.7: Remove `Nullable`

As the major stakeholders, I would appreciate opinions from
- @bergmark 
- @ocharles 
- @hesselink 
- @dbp
- @c-lewis 
- @coubeatczech 
- @tstat

**Comment by bergmark** [comment]
I think this would be pretty hairy for us to work around...

As for `.==`: We solved this by not allowing comparisons of `Nullable` and using a `.==?` for them instead.

I also don't see what's nonsensical about `Nullable (Nullable a)`. You can't distinguish between `Nullable Null` and `Null` but I thought that was the reason `Maybe` isn't used for these columns. We do the same thing in Fay.

**Comment by tomjaguarpaw** [comment]
> I think this would be pretty hairy for us to work around...

Can you say more about why?

> We solved this by not allowing comparisons of `Nullable`

How did you acheive that?

**Comment by tomjaguarpaw** [comment]
I see you achieved it thus: `(.==) :: ShowConstant a => Column a -> Column a -> Column Bool`

**Comment by bergmark** [comment]
> Can you say more about why?

Sure,

We currently only use two type aliases, one for the haskell values and one opaleye, e.g.:
```
data PersonP a b = Person { name :: a, gender :: b }
type PersonO = Person String (Nullable Gender)
type PersonH = Person String (Maybe Gender)
```
A normal query for a person gives us `To Column PersonO:` with `Nullable` already inside the type alias, we'd have to pull that out somehow and the `To` type family may not be usable anymore.

A left join gives us `MapTo Column (OtherType, To Nullable PersonO) = (To Column OtherType, To Column (To Nullable PersonO))`, this would also need to change somehow.

**Comment by tomjaguarpaw** [comment]
I can see that would be problematic.

**Comment by bergmark** [comment]
Since Opaleye has `PGOrd` already we could have `PGEq a :: Column a -> Column a -> PGBool`. We went through this migration and It was a good thing to do. Type errors were easy to fix, first consider whether the postgres semantics for null equality is what we want and then change to `.==?`, possibly calling `nullable` on one of the operands.

**Comment by k0001** [comment]
I agree with @tomjaguarpaw that `Column a` vs `NullableColumn a` is a better represenation than `Column a` vs `Collumn (Nullable a)`.

As an additional data point, I'm currently working on a thin layer on top of Opaleye to provide a more robust API in scenarios like this one: I'm trying to make the distinction between `Column (Nullable a)` and `Column a` as explicit as possible, yet allow some operations (say, something like `.==`) to work correctly both in nullable and not nullable columns, but there are many cases where implementing instances for those overloaded operations is inconvenient or impossible because of overlapping instances between `Column (Nullable a)` and `Column a`.

**Comment by tomjaguarpaw** [comment]
Renzo, I suggest you go ahead and define `newtype NullableColumn a = NullableColumn (Column a)` somewhere in your own library and report back how it goes!  You may have to write a bunch of boilerplate to convert to existing Haskell functions but it's probably worthwhile nonetheless.

**Comment by k0001** [comment]
Good idea Tom, I might just do that.  Thanks.
Il 10/set/2015 06:27 PM, "tomjaguarpaw" <notifications@github.com> ha
scritto:

**Comment by k0001** [comment]
I wanted to mention that in my `opaleye-sot`, the thin API layer on top of `opaleye` I'm working on, I've made the switch to this representation. I'm quite happy with the results so far, having this explicit distinction between what can and what can't be nullable brings the benefits of `Int` vs `Maybe Int` to the DSL. This distinction makes it much easier to reason about what can and can't be NULL.

To avoid any confusions between Opeleye's current `Column` and this other representation that forbids having a `Column (Nullable a)`, I've temporarily renamed the `Column` proposed here to `Kol`, and the `NullableColumn` proposed here to `Koln`. You can see my work on `Kol` and `Koln` in https://github.com/k0001/opaleye-sot/blob/master/src/lib/Opaleye/SOT/Internal.hs

Keep in mind that said code is a work in process, and that in that very same module a lot of type-level programming is going on for purposes unrelated to the implementation of `Kol` and `Koln`, so you can ignore that. If you are interested in learning about my goals with `opaleye-sot`, you can take a look at some of the examples in the tutorial I'm working on: https://github.com/k0001/opaleye-sot/blob/master/src/lib/Opaleye/SOT/Tutorial.hs

**Comment by tomjaguarpaw** [comment]
Thanks for the progress report.  Glad to hear it's working out well.

**Comment by ocharles** [comment]
I think there's one more oddity that you might want to consider here. Functions like `restrict` actually have domain `NullableColumn Bool` rather than just `Column Bool`. However, I believe `NULL` *does* have a special meaning here - it's the same as `false`. Likewise for `LEFT JOIN`.

**Comment by tomjaguarpaw** [comment]
I'm not so worried about that.  Anyone who wants to restrict or join on a genuinely nullable condition is mad.

**Comment by ocharles** [comment]
Err, no they aren't.

```sql
SELECT * X LEFT JOIN Y ON (X.A = Y.A) LEFT JOIN Z ON (Y.A = Z.A)
```

`Y.A` can be `null`.

**Comment by tomjaguarpaw** [comment]
That's something else. The return value of the equality cannot be `NULL`.

**Comment by tomjaguarpaw** [comment]
Actually I rescind my claim of madness in the join case.  Due to the use of indexes I can see that you might want to join on nullable conditions.  I don't think you ever want to restrict on one though.

**Comment by ocharles** [comment]
Surely any arguments for indexes applies equally well for `WHERE` clauses then. But I have no examples there, but I do have examples with joins (the above is just an alpha-renaming & simplification of a real query we have).

**Comment by tomjaguarpaw** [comment]
OK you convinced me.  I'm the mad one :)

**Comment by xnmp** [comment]
Is there any progress on this at all, and would it solve the problem of having to to manually type annotate left joins? Doing that (and also having to type annotate any usage of `runQuery`) is by far the single most annoying thing about Opaleye for me. 

the type of left join could just change to 

```haskell
leftJoin :: (D.Default U.Unpackspec (Column a) (Column a),
              D.Default U.Unpackspec (NullableColumn a) (NullableColumn a),
          => Query (Column a)  -- ^ Left query
          -> Query (NullableColumn a)  -- ^ Right query
          -> (((Column a), (NullableColumn a)) -> Column T.PGBool) -- ^ Condition on which to join
          -> Query ((Column a), nullable(NullableColumn a)) -- ^ Left join
```

With of course appropriate modifications to dealing with records and tuples. (```Column User ... = User Column ...```), and ```Column (a,b) = (Column a, Column b)```.
Since I only write select queries I'm sacrificing type safety so that I don't have to annotate, with:

```
null2 :: Column a
null2 = O.Column (HPQ.ConstExpr HPQ.NullLit)

isNull2 :: Column a -> Column Bool
isNull2 = O.unOp HPQ.OpIsNull
```

**Comment by tstat** [comment]
@bergmark @ocharles @hesselink @dbp @c-lewis back in June I implemented the change from `Column (Nullable a)` to `Column (n :: Nullability) a` in this PR: https://github.com/tomjaguarpaw/haskell-opaleye/pull/305. I haven't really revisited this issue (or the opaleye codebase) since then, so I may need to refresh myself on all of the possible implications of this change, but it seems like an obvious improvement to me for a couple of reasons.

1.  `Nullable (Nullable a)` and `Nullable a` are equivalent in that the possible return values from postgres are still `a`, or `NULL`, that is, it is impossible to get a `Just Nothing` back from such a query, but they are distinct at the type level.
2. With the `Nullable a` construction we cannot represent arbitrary non-nullable types parametrically, This causes the type restrictions in some functions to be nonsensical. For instance, consider the functions for manipulating JSON. I have pasted one of these functions below:
```haskell
(.#>) :: (PGIsJson a)
      => Column (C.Nullable a) -- ^
      -> Column (T.PGArray T.PGText) -- ^ path
      -> Column (C.Nullable a)
```
We make the first argument nullable so that we can ensure the return column is a nullable value of type `a` without running the risk of returning a nested nullable. Restricting the type of the first argument to be nullable is not actually meaningful, it is just a trick to make sure that the return value is nullable, but not a nested nullable. As another example of this same issue, I introduced `joinNullable :: Nullable (Nullable a) -> Nullable a` in my PR for adding indexing on arrays, as `index :: (C.PGIntegral n) => Column (T.PGArray a) -> Column n -> Column (C.Nullable a)` would return a nested `Nullable` if the input array has nullable values. So we are having to do some type dancing that is not actually meaningful. The `Column (n :: Nullability) a` approach solves this issue since we can represent arbitrary non-nullable types parametrically. https://github.com/tomjaguarpaw/haskell-opaleye/pull/305/commits/d2c94d1d03fa8ee6d74a5525e59d44118fe4583a#diff-7eeb6cd9196c52b6b8e2d6da8c6a0265L261

We could also solve issues like https://github.com/tomjaguarpaw/haskell-opaleye/issues/91 with this approach.

So to summarize: I don't see how `Nullable a` gives us any distinct information when compared to a `Column (n :: Nullability) a` approach, yet we are introducing arbitrary type restrictions and doing small type dances to satisfy them. So I propose we move to a `Column (n :: Nullability) a` approach.

How do you all feel about this proposal?

**Comment by tomjaguarpaw** [comment]
> How do you all feel about this proposal?

My feelings on the matter are exactly the same as before:

> The idea is sound[1] and basically The Right Thing™ but it's an *enormous* change and the practical benefit over the status quo is small. This means that the time before I'll make this change is measured in months and, to be quite realistic, I may never get round to it at all.
>
> -- https://github.com/tomjaguarpaw/haskell-opaleye/pull/305#issuecomment-315025293

To be clear, "enormous" in the above means in terms of the breakage it will cause clients, rather than in the amount of work it would take to implement.

I think it's worth talking through some of the issues and seeing how people feel about them.  For example *every* column would have to be tagged with its nullability, not just nullable ones.  We'll see
`Column NonNullable PGType` everywhere, instead of just `Column PGType`.  We could shorten this to `Column NN PGType`, but it's still more text than before.  Personally I think it's easily a price worth paying, but does everyone else?

**Comment by tstat** [comment]
> I think it's worth talking through some of the issues and seeing how people feel about them.

Yes, I agree; that is certainly my intention with this post. I know that @bergmark and @ocharles had issues with this proposal when you initially made it. I suspect some of these potential problems might be alleviated since this issue was originally proposing separate constructors.

> For example every column would have to be tagged with its nullability, not just nullable ones. We'll see
Column NonNullable PGType everywhere, instead of just Column PGType. We could shorten this to Column NN PGType, but it's still more text than before. Personally I think it's easily a price worth paying, but does everyone else?

I think the specific issue you raise of verbosity is a legitimate as well, especially with new type variables to type classes, however, I will point out that we can shorten it beyond `Column NN PGType` and simply use `Column PGType` and `NullableColumn PGType` by declaring a couple of type aliases.

https://github.com/tomjaguarpaw/haskell-opaleye/pull/305/commits/0ec5bcd48087987cfb596555a36f762e07cc65ba#diff-a04b9a32b4efa5892b2d9309cb17af57R15

**Comment by tomjaguarpaw** [comment]
>  I know that @bergmark and @ocharles had issues with this proposal when you initially made it.

I think the idea of using a type parameter instead of two separate type constructors probably addresses their points.

**Comment by tomjaguarpaw** [comment]
Fixed by introduction of `Field`.
